x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -337,107 +348,163 @@ private RestoreSession openSession(String repositoryName, Client remoteClient, S
+                        }
+                        final long maxConcurrentFileChunks = ccrSettings.getMaxConcurrentFileChunks();
+                        if (remainingFiles.isEmpty() == false && inFlightRequests.size() < maxConcurrentFileChunks) {
+                            for (int i = 0; i < maxConcurrentFileChunks; i++) {
+                                if (remainingFiles.isEmpty()) {
+                                    break;
+                                }
+                                inFlightRequests.push(remainingFiles.pop());
+                            }
+                        }
+                        fileToRecover = inFlightRequests.removeFirst();
+                        inFlightRequests.addLast(fileToRecover);
+                        waitForOp = requestSeqIdTracker.getMaxSeqNo() + 1 - inFlightRequests.size();
+                    }
+                    try {
+                        requestSeqIdTracker.waitForOpsToComplete(waitForOp);
+                        synchronized (mutex) {
+                            // if file has been removed in the mean-while, it means that restore of this file completed, so start working
+                            // on the next one
+                            if (inFlightRequests.contains(fileToRecover) == false) {
+                                continue;
+                            }
+                        }
+                        final int bytesRequested = Math.toIntExact(Math.min(ccrSettings.getChunkSize().getBytes(), fileToRecover.length()));
+                        final GetCcrRestoreFileChunkRequest request =
+                            new GetCcrRestoreFileChunkRequest(node, sessionUUID, fileToRecover.name(), bytesRequested);
+                        logger.trace("[{}] [{}] fetching chunk for file [{}]", shardId, snapshotId, fileToRecover.name());
+
+                        final long requestSeqId = requestSeqIdTracker.generateSeqNo();
+                        try {
+                            remoteClient.execute(GetCcrRestoreFileChunkAction.INSTANCE, request,
+                                ActionListener.wrap(
+                                    r -> threadPool.generic().execute(new AbstractRunnable() {
+                                        @Override
+                                        public void onFailure(Exception e) {
+                                            error.compareAndSet(null, Tuple.tuple(fileToRecover.metadata(), e));
+                                            requestSeqIdTracker.markSeqNoAsCompleted(requestSeqId);
+                                        }
+
+                                        @Override
+                                        protected void doRun() throws Exception {
+                                            final int actualChunkSize = r.getChunk().length();
+                                            final long nanosPaused = ccrSettings.getRateLimiter().maybePause(actualChunkSize);
```

Comment:
It feels like we should throttle after writing the file chunk? Since I think the rate limiting activity more about network activity, we might as well spend some time writing to disk before determining if we want to pause.

Reply:
I've chosen to use the same way we do for peer recovery, to keep things consistent. I'm fine if we want to follow-up on this with a change to both.